File: markdowner-core/src/main/java/org/jesperancinha/parser/markdowner/badges/parser/BadgeParser.java
package org.jesperancinha.parser.markdowner.badges.parser;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.jesperancinha.parser.markdowner.badges.model.*;

import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Slf4j
public class BadgeParser {

    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static final String GENERIC_REGEX = "a-zA-Z0-9\\/\\.\\]\\?\\=\\-\\&\\%%\\;\\_\\#\\:\\@\\\"\\ ";
    private static final String EMOJI_REGEX =
            "\uD83D\uDD8B\u00a9\u00ae\u2000-\u3300\ud83c\ud000-\udfff\ud83d\ud000-\udfff\ud83e\ud000-\udfff\ufe0f\uD83C-\uD83D\uDCDA-\uDFE2";
    public static final String FULL_REGEX = GENERIC_REGEX.concat(EMOJI_REGEX);
    private static final String BADGE_REGEX =
            "(\\[!\\[%s]\\(http[s]*:\\/\\/%s[" + FULL_REGEX + "]*\\)]\\((http[s]*:\\/\\/)*[" + FULL_REGEX + "]*\\))";


    private static final Pattern NOT_ACCEPTED_REGEX =
            Pattern.compile("color=(?!(informational)).");
    public static final Map<String, BadgeType> badgeTypes = parseBadgeTypes();

    private static Map<String, BadgeType> parseBadgeTypes() {
        try {
            return Arrays.stream(objectMapper.readValue(BadgeParser.class.getResourceAsStream("/jeorg.badges.types.json")
                            , BadgeType[].class))
                    .collect(Collectors.toMap(BadgeType::getType, badgeType -> badgeType));
        } catch (IOException e) {
            log.error("Error!", e);
            System.exit(1);
        }
        return null;
    }

    public static final Map<BadgeType, BadgeSettingGroup> badgeSettingGroups = parseSettings();

    public static Map<BadgeType, BadgeGroup> parse(final String readmeText) {
        return badgeSettingGroups
                .values()
                .stream()
                .filter(Objects::nonNull)
                .collect(Collectors.toMap(BadgeSettingGroup::getBadgeType, badgeSettingGroup -> {
                            final Map<Pattern, Badge> allBadges = badgeSettingGroup
                                    .getBadgeSettingList()
                                    .stream()
                                    .filter(Objects::nonNull)
                                    .collect(HashMap::new, (map, badgeSetting) -> {
                                        final Matcher matcher = badgeSetting.getPattern().matcher(readmeText);
                                        if (matcher.find()) {
                                            final String badgeText = matcher.group(0);
                                            final Matcher matcher1 = NOT_ACCEPTED_REGEX.matcher(badgeText);
                                            final String linkPrefix = badgeSetting.getLinkPrefix();
                                            if (matcher1.find() || linkPrefix != null && !badgeText.contains(linkPrefix)) {
                                                map.put(badgeSetting.getPattern(), null);
                                            } else {
                                                map.put(badgeSetting.getPattern(), Badge.builder()
                                                        .badgeText(badgeText)
                                                        .title(badgeSetting.getTitle())
                                                        .build());
                                            }
                                        } else {
                                            map.put(badgeSetting.getPattern(), null);
                                        }
                                    }, HashMap::putAll);


                            return BadgeGroup
                                    .builder()
                                    .badgeType(badgeSettingGroup.getBadgeType())
                                    .badgeHashMap(allBadges)
                                    .build();
                        }
                ));
    }


    static Map<BadgeType, BadgeSettingGroup> parseSettings() {
        return badgeTypes.values().stream()
                .collect(Collectors.toMap(badgeType -> badgeType, badgeType -> {
                    try {
                        final var badgeSettingList = Arrays.stream(objectMapper.readValue(
                                BadgeParser.class.getResourceAsStream("/".concat(badgeType.getBadgeFile())),
                                BadgeSetting[].class)).collect(Collectors.toList());
                        return BadgeSettingGroup
                                .builder()
                                .badgeType(badgeType)
                                .badgeSettingList(
                                        badgeSettingList
                                                .stream().map(badgeSetting ->
                                                        BadgePattern.builder()
                                                                .title(badgeSetting.getTitle())
                                                                .pattern(Pattern.compile(
                                                                        String.format(BADGE_REGEX,
                                                                                badgeSetting.getBadge(),
                                                                                badgeSetting.getCodePrefix()
                                                                                        .replace(".", "\\.")
                                                                                        .replace("/", "\\/")
                                                                        )))
                                                                .linkPrefix(badgeSetting.getLinkPrefix())
                                                                .build())
                                                .collect(Collectors.toList())
                                )
                                .build();
                    } catch (IOException e) {
                        log.error("Error", e);
                    }
                    return null;
                }));
    }
}

File: markdowner-core/src/main/kotlin/org/jesperancinha/parser/markdowner/badges/parser/BadgeParser.kt
package org.jesperancinha.parser.markdowner.badges.parser

import com.fasterxml.jackson.databind.ObjectMapper
import org.jesperancinha.parser.markdowner.badges.model.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.*
import java.util.regex.Pattern
import kotlin.system.exitProcess

class BadgeParser {

    companion object{
        private const val GENERIC_REGEX = "a-zA-Z0-9\\/\\.\\]\\?\\=\\-\\&\\%%\\;\\_\\#\\:\\@\\\"\\ "
        private const val EMOJI_REGEX = ".{0,4}"
        const val FULL_REGEX = GENERIC_REGEX + EMOJI_REGEX
        private const val BADGE_REGEX =
            "(\\[!\\[%s]\\(http[s]*:\\/\\/%s[$GENERIC_REGEX]*$EMOJI_REGEX[$GENERIC_REGEX]*\\)]\\((http[s]*:\\/\\/)*[$GENERIC_REGEX]*\\))"
        private val NOT_ACCEPTED_REGEX = Pattern.compile("color=(?!(informational)).")
        private val objectMapper = ObjectMapper()
        private val logger: Logger = LoggerFactory.getLogger(BadgeParser::class.java)

        private val badgeTypes = parseBadgeTypes()
        private fun parseBadgeTypes(): Map<String, BadgeType> = try {
            objectMapper.readValue(
                BadgeParser::class.java.getResourceAsStream("/jeorg.badges.types.json"),
                Array<BadgeType>::class.java
            ).associateBy { it.type }
        } catch (e: IOException) {
            logger.error("Error!", e)
            exitProcess(1)
        }

        val badgeSettingGroups = parseSettings()

        fun parse(readmeText: String): Map<BadgeType, BadgeGroup> {
            return badgeSettingGroups.values
                .filterNotNull().associate { badgeSettingGroup ->
                    val allBadges = badgeSettingGroup.badgeSettingList.associate { badgePattern ->
                        badgePattern.pattern to run {
                            val matcher = badgePattern.pattern.matcher(readmeText)
                            if (matcher.find()) {
                                val badgeText = matcher.group(0)
                                val matcher1 = NOT_ACCEPTED_REGEX.matcher(badgeText)
                                val linkPrefix = badgePattern.linkPrefix
                                if (matcher1.find() || linkPrefix != null && !badgeText.contains(linkPrefix)) {
                                    null
                                } else {
                                    Badge(
                                        badgeText = badgeText,
                                        title = badgePattern.title
                                    )
                                }
                            } else {
                                null
                            }
                        }
                    }
                    badgeSettingGroup.badgeType to BadgeGroup(
                        badgeType = badgeSettingGroup.badgeType,
                        badgeHashMap = allBadges
                    )
                }
        }

        fun parseSettings(): Map<BadgeType, BadgeSettingGroup?> =
            badgeTypes.values.associateWith { badgeType ->
                try {
                    val badgeSettingList = objectMapper.readValue(
                        BadgeParser::class.java.getResourceAsStream("/" + badgeType.badgeFile),
                        Array<BadgeSetting>::class.java
                    )
                    BadgeSettingGroup(
                        badgeType = badgeType,
                        badgeSettingList = badgeSettingList.map { badgeSetting ->
                            BadgePattern(
                                title = badgeSetting.title,
                                pattern = Pattern.compile(
                                    String.format(
                                        BADGE_REGEX,
                                        badgeSetting.badge,
                                        badgeSetting.codePrefix
                                            .replace(".", "\\.")
                                            .replace("/", "\\/")
                                    )
                                ),
                                linkPrefix = badgeSetting.linkPrefix
                            )
                        }
                    )
                } catch (e: IOException) {
                    logger.error("Error", e)
                    null
                }
            }

    }
}